### ISSUE TITLE
Applying more than one reference set to SingleR analysis

### ISSUE THREAD

**Issue by erikfel97** [LTLA/SingleR#54]
Hi all,
In order to subject my data to a bigger reference set, I would like to combine some reference sets together: HumanPrimaryCellAtlasData(), BlueprintEncodeData() and MonacoImmuneData(). Is there an option to do this?

Best Regards,
Erik

**Comment by LTLA** [comment]
To answer your immediate technical question: just `cbind` the count matrices together. You'll need to take the intersection of common genes, though. For example:

```r
x1 <- HumanPrimaryCellAtlasData()
x2 <- BlueprintEncodeData()
x3 <- MonacoImmuneData()

# Taking the intersection of all genes and combining them.
common <- Reduce(intersect, list(rownames(x1), rownames(x2), rownames(x3)))
combined <- cbind(x1[common,], x2[common,], x3[common,])
```

This could then be fed into `SingleR` as a single reference, and it probably wouldn't do too badly. However, to be truly effective, some more work is required. You will notice, for example, that B cells are split into "B cells", "B_cells" and "B-cells". (Hey, don't blame us; this was how the annotation was provided.) We would thus like to synchronize the labels before combining them, and this can be guided using the `matchReferences` function. For example, to match references between `x1` and `x2`:

```r
match.2.to.1 <- matchReferences(x2, x1, x2$label.main, x1$label.main)
new.labels.for.2 <- colnames(match.2.to.1)[max.col(match.2.to.1)]
data.frame(old=rownames(match.2.to.1), new=new.labels.for.2) # kind of makes sense.
x2$label.main <- new.labels.for.2[match(x2$label.main, rownames(match.2.to.1))]
```

You might argue that this is too complicated, and that's true enough, but it's challenging to synchronize annotation across different datasets; some level of expert inspection **is** required. For example, the data.frame above has a mapping of smooth muscle cells to platelets that would have to be manually removed, and the user has to be able to spot that and intervene.

Anyway, should you complete synchronization of all labels, you can proceed to using `SingleR()` as described above. But this leads to the next question... why do you want to assemble a combined reference? Is there a specific problem that you cannot solve with a single reference? I would be inclined to annotate with each reference separately and then combine the results, e.g., by taking the label with the highest score for each cell across references.

**Comment by LTLA** [comment]
And on that note, it would be good if someone could write a `combineResults` function that takes assignments of one dataset to several different references and combines them into a single set of assignments. I'm thinking of just taking the label with the highest first-round score, which should be good enough (assuming the fine-tuning doesn't pick a label that's too far from the first round).

I will handle the wrapper around `matchReferences` to make it a bit more user-friendly.

**Comment by LTLA** [comment]
This is now available in v1.0.1 by passing a list of references to `SingleR()`. See `?combineResults` for more theoretical details.
